Add dumping for NamedConstraintId and shorten untagged id printing

danakj opened the pull request:
Adds support to the `dump` debugger command for named constraint ids, which are printed as `constraint<number>`. While doing so, we print whether the `constraint` is complete or not, and add the same to `interface` to match.

And we noticed that the printing of name and name scope ids, which are not tagged, are very verbose by adding 7 `0`s to them for no reason. So make the dump output easier to read by dropping 0 prefixes.

Before:
```
name_scope00000000: {inst: inst0000000E, parent_scope: name_scope<none>, has_error: false, extended_scopes: [], names: {name00000000: inst6000000F, name00000001: inst60000011}} {kind: Namespace, arg0: name_scope00000000, arg1: inst<none>, type: type(inst(NamespaceType))} `package`
```

After:
```
name_scope0: {inst: instE, parent_scope: name_scope<none>, has_error: false, extended_scopes: [], names: {name0: inst6000000F, name1: inst60000011}} {kind: Namespace, arg0: name_scope0, arg1: inst<none>, type: type(inst(NamespaceType))} `package`
```

danakj committed:
dump-constraint

zygoloid reviewed (commented):
sem_ir/dump changes LGTM

I'd like @dwblaikie to look at the change to `IdBase::Print` -- I think we probably did this to make the printing consistent across tagged IDs with a tag of 0 and those with a nonzero tag, but I don't have the relevant state there.

dwblaikie commented:
> sem_ir/dump changes LGTM
> 
> I'd like @dwblaikie to look at the change to `IdBase::Print` -- I think we probably did this to make the printing consistent across tagged IDs with a tag of 0 and those with a nonzero tag, but I don't have the relevant state there.

We switched the non-indexed IDs to hex dumping in f64b08863a23bcfb3e40f0c5ed55413559bd2948 addressing feedback in https://github.com/carbon-language/carbon-lang/pull/6215#discussion_r2430180953

But I don't think there was a particular discussion about how we should print hex - I guess one motivation to print it in a fixed width is to make it clearer it's different from the decimal cases? We could print the 0x prefix to do that instead & then print it smaller? 

seeing `thing42` and not knowing whether the 42 is decimal or hex seems a bit unfortunate - I guess not too much of an issue for you @danakj because you copy/paste these, never treat them as raw numbers in debugger expressions, etc?

danakj commented:
> > sem_ir/dump changes LGTM
> > I'd like @dwblaikie to look at the change to `IdBase::Print` -- I think we probably did this to make the printing consistent across tagged IDs with a tag of 0 and those with a nonzero tag, but I don't have the relevant state there.
> 
> We switched the non-indexed IDs to hex dumping in [f64b088](https://github.com/carbon-language/carbon-lang/commit/f64b08863a23bcfb3e40f0c5ed55413559bd2948) addressing feedback in [#6215 (comment)](https://github.com/carbon-language/carbon-lang/pull/6215#discussion_r2430180953)
> 
> But I don't think there was a particular discussion about how we should print hex - I guess one motivation to print it in a fixed width is to make it clearer it's different from the decimal cases? We could print the 0x prefix to do that instead & then print it smaller?
> 
> seeing `thing42` and not knowing whether the 42 is decimal or hex seems a bit unfortunate - I guess not too much of an issue for you @danakj because you copy/paste these, never treat them as raw numbers in debugger expressions, etc?

Yeah I thought about that but they are _all_ hex now, so it seems like it's not an issue anymore. `thing6000001` doesn't say hex anywhere either.

danakj committed:
Merge remote-tracking branch 'origin' into dump-constraint

dwblaikie reviewed (approved):
> Yeah I thought about that but they are _all_ hex now, so it seems like it's not an issue anymore. `thing6000001` doesn't say hex anywhere either.

For small test cases at least, it's pretty easy to spot the difference between the hex and decimal values - and there are some decimal values (The "Index" Ids, like `call_paramN` and `argN`, etc).

But I think you're the one making the most of these values - so if it works for you, let's do it!

danakj commented:
> > Yeah I thought about that but they are _all_ hex now, so it seems like it's not an issue anymore. `thing6000001` doesn't say hex anywhere either.
> 
> For small test cases at least, it's pretty easy to spot the difference between the hex and decimal values - and there are some decimal values (The "Index" Ids, like `call_paramN` and `argN`, etc).
> 
> But I think you're the one making the most of these values - so if it works for you, let's do it!

Ok sounds good. If we hear anyone being confused, let's put it back.
